### PR TITLE
Trim file type bits from mode header

### LINKED ISSUE #77096: Invalid UnixFileMode value when using the TarFile API
### Description

I'm trying to exact a tar file into a folder, and the tar file may contains folders and files, when I try to use the `TarFile.ExtractToDirectory` API, I got the error like follows, maybe I'm wrong on using the API

```
Exception when invoke command System.ArgumentException: Invalid UnixFileMode value. (Parameter 'unixCreateMode')
   at System.IO.Directory.CreateDirectoryCore(String path, UnixFileMode unixCreateMode)
   at System.Formats.Tar.TarHelpers.CreateDirectory(String fullPath, Nullable`1 mode, SortedDictionary`2 pendingModes)
   at System.Formats.Tar.TarEntry.ExtractRelativeToDirectoryAsync(String destinationDirectoryPath, Boolean overwrite, SortedDictionary`2 pendingModes, CancellationToken cancellationToken)
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternalAsync(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen, CancellationToken cancellationToken)
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternalAsync(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen, CancellationToken cancellationToken)
```

### Reproduction Steps

``` c#
var tmpDirPath = Path.Combine(TmpPath, Guid.NewGuid().ToString("N"));
await var fs = File.OpenRead(@"xx.tar.gz");
await using var decompressStream = new GZipStream(fs, CompressionMode.Decompress);
await TarFile.ExtractToDirectoryAsync(decompressStream, tmpDirPath, true);
```

### Expected behavior

Work as normal

### Actual behavior

Exception as before

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 7 RC2
CentOS 7

### Other information

Works well on Windows but not on CentOS

### PR DESCRIPTION
**Pull request by am11**
Fixes https://github.com/dotnet/runtime/issues/77096.

<del>Depends on https://github.com/dotnet/runtime-assets/pull/279.</del>

### CODE REVIEW

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (tmds) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:141]: Can you update the comment so it reflects we're returning the `UnixFileMode` bits?

There's a `const` for these, but you probably can't access it here:

https://github.com/dotnet/runtime/blob/3f600a1afbb19034ec533b3b145ddb5b741623f0/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.cs#L8

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (tmds) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:144]: Same here. Can the comment be updated to reflect we're checking for valid values of `UnixFileMode`?
I've written this check elsewhere as `(value & ~FileSystem.ValidUnixFileModes) != 0)`.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (am11) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:141]: Sure, I will define this constant in S.F.Tar to keep the uniform style.

**Comment by tmds** [comment]
> Fixes https://github.com/dotnet/runtime/pull/77107.

I think you meant to link to: https://github.com/dotnet/runtime/issues/77096

**Comment by am11** [comment]
> I think you meant to link to: #77096

Yes, fixed. Too many tabs and PRs are opened. 😅

**Comment by adamsitnik** [comment]
> Depends on https://github.com/dotnet/runtime-assets/pull/279.

Merged ;)

**Review by tmds — APPROVED** [review]
lgtm, thanks!

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you very much for your help @am11 !

**Comment by danmoseley** [comment]
@am11 I moved to the ASPNET team so it's  their call. Test coverage doesn't need any special approvals to back port though. 7.0 branch might be closed for a bit for GA.

**Comment by danmoseley** [comment]
Oh this is a product change. It would need all the regular process. The bar is roughly the servicing bar right now. It would be up to the folks in this repo though.

**Comment by am11** [comment]
@danmoseley this is the fix for the issue https://github.com/dotnet/runtime/issues/77096 in which you were coordinating with the author. Seeing other fixes being backported to 7.0, so there may still be some time.

**Comment by danmoseley** [comment]
Yup, but this is @adamsitnik / @jeffhandley  area now.

The test failure is another hit on https://github.com/dotnet/runtime/issues/77087
0:014> k
 # ChildEBP RetAddr      
00 (Inline) --------     coreclr!WKS::gc_heap::mark_array_marked+0x1a [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 8471] 
01 (Inline) --------     coreclr!WKS::gc_heap::background_mark1+0x1a [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 23743] 
02 1273f8e0 738482f8     coreclr!WKS::gc_heap::background_mark_simple+0x1b [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 25142] 
03 (Inline) --------     coreclr!WKS::gc_heap::background_mark_object+0x22 [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 25162] 
04 1273f8e0 73847b01     coreclr!WKS::gc_heap::revisit_written_page+0x3c8 [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 35631] 
05 1273f938 73885ffc     coreclr!WKS::gc_heap::revisit_written_pages+0xf1 [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 35832] 
06 1273f980 7383195b     coreclr!WKS::gc_heap::background_mark_phase+0x1b2 [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 35093] 
07 1273f9b8 738a0d4c     coreclr!WKS::gc_heap::gc1+0x49e [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 21184] 
08 1273f9d0 738a0cef     coreclr!WKS::gc_heap::bgc_thread_function+0x5a [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 36283] 
09 1273f9f8 738b8e50     coreclr!WKS::gc_heap::bgc_thread_stub+0x1f [D:\a\_work\1\s\src\coreclr\gc\gc.cpp @ 34225] 
0a 1273f9f8 75a46709     coreclr!<lambda_410a0441ee61a5b8ed4f18f53fb141bc>::operator()+0x70 [D:\a\_work\1\s\src\coreclr\vm\gcenv.ee.cpp @ 1415]

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:146]: nit: There's another reference to `4095` on the tests which we could change to a `const UnixFileMode ValidUnixFileModes`.
https://github.com/dotnet/runtime/blob/ccd9d169b0189d0503ec885bf86e9f026cb621a8/src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.File.Base.Unix.cs#L38

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:142]: I wonder if not using the setter everywhere is correct. it seems somewhat inconsistent and we could've thrown earlier.

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs:101]: I would also include a version of this test using a compressed archive since that's what the customer reported. Also, add the sync test. But this is up to you, not a big deal tbh.

**Review by jozkee — APPROVED** [review]
LGTM!

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (am11) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:146]: Currently we don't have access to the helpers in the tests.

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs
Comment (am11) [reply src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs:101]: That will require setting up token to fetch gzipped tar archive from docker api and posting a PR like https://github.com/dotnet/runtime-assets/pull/279. I don't have access to those tokens (and don't think it is super important for 7.0 minimal fix).

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (am11) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:142]: Good point, I am not sure if this is intentional. I didn't wanted to go for larger refactoring at this point. There is also an existing helper: https://github.com/dotnet/runtime/blob/88aea31185824d30cb1a2edcfe2d0472000d9e22/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs#L55, but for some reason we don't use it in the path OP was hitting the exception.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (Jozkee) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:146]: I was suggesting _copying_ the `const` to the tests, not making the one on src accessible to tests. Is just a nit, so you can disregard it.

**Comment by Jozkee** [comment]
Thanks.

**Comment by am11** [comment]
@Jozkee, thanks. Can you please backport it to 7.0? This is matching the behavior of other tar reader implementations (`tar xf docker-hello-world.tar` on mac or linux works).

**Comment by jeffhandley** [comment]
@am11 I'm going to initiate the backport. This will put the fix into consideration for a servicing release--most of the other backports you've observed have also been for the first servicing release. Thanks for this fix!

**Pull request merged**
